src/Compilers/CSharp/Portable/Binder/Semantics/BestTypeInferrer.cs

```diff
@@ -13,62 +13,22 @@ internal static class BestTypeInferrer
         public static NullableAnnotation GetNullableAnnotation(TypeSymbol bestType, ArrayBuilder<TypeSymbolWithAnnotations> types)
         {
             bool bestTypeIsPossiblyNullableReferenceTypeTypeParameter = bestType.IsPossiblyNullableReferenceTypeTypeParameter();
-            NullableAnnotation? result = null;
+            NullableAnnotation result = NullableAnnotation.NotAnnotated;
             foreach (var type in types)
             {
                 if (type.IsNull)
                 {
                     // https://github.com/dotnet/roslyn/issues/27961 Should ignore untyped
                     // expressions such as unbound lambdas and typeless tuples.
-                    result = NullableAnnotation.Nullable;
+                    result = NullableAnnotation.Annotated;
                     continue;
                 }
 
-                if (!type.IsReferenceType && !type.TypeSymbol.IsPossiblyNullableReferenceTypeTypeParameter())
-                {
-                    return NullableAnnotation.Unknown;
-                }
-
-                NullableAnnotation nullableAnnotation;
-
-                if (type.IsPossiblyNullableReferenceTypeTypeParameter() && !bestTypeIsPossiblyNullableReferenceTypeTypeParameter)
-                {
-                    nullableAnnotation = NullableAnnotation.Nullable;
-                }
-                else
-                {
-                    nullableAnnotation = type.NullableAnnotation;
-                }
-
-                if (nullableAnnotation == NullableAnnotation.Unknown)
-                {
-                    if (result?.IsAnyNotNullable() != false)
-                    {
-                        result = NullableAnnotation.Unknown;
-                    }
-                }
-                else if (nullableAnnotation.IsAnyNullable())
-                {
-                    if (result?.IsAnyNullable() != true)
-                    {
-                        result = nullableAnnotation;
-                    }
-                    else if (result != nullableAnnotation)
-                    {
-                        result = NullableAnnotation.Annotated;
-                    }
-                }
-                else if (result == null)
-                {
-                    result = nullableAnnotation;
-                }
-                else if (result.GetValueOrDefault() == NullableAnnotation.NotNullable && nullableAnnotation == NullableAnnotation.NotAnnotated)
-                {
-                    result = NullableAnnotation.NotAnnotated;
-                }
+                // This uses the covariant merging rules.
+                result = result.JoinForFixingLowerBounds(type.AsSpeakable().NullableAnnotation);
```

Comment:
I believe this function assumes that we are operating on the same types, but I think the ```types``` list can contain different types today.

Reply:
If you look at the intermediate PRs, I added a harness to this method before changing it. So I know that the only thing that changed is the range of allowed inputs.

Reply:
> If you look at the intermediate PRs, I added a harness to this method before changing it. So I know that the only thing that changed is the range of allowed inputs.

I am not sure how to interpret this response. Given the current situation, it looks like the use of ```JoinForFixingLowerBounds``` is unjustified unless we can assert that ```types``` contains the same types ignoring nullability.

---
In reply to: [244027745](https://github.com/dotnet/roslyn/pull/31813#discussion_r244027745) [](ancestors = 244027745)

Reply:
Looking at the two callers (array initializers and lambdas), I think the callers need to be fixed so that we get here with a single type (ie. after all the elements have been converted to a best type). Otherwise, we miss nullability effects of user-defined conversions.

---
In reply to: [244031641](https://github.com/dotnet/roslyn/pull/31813#discussion_r244031641) [](ancestors = 244031641,244029088,244027745)